aj/add layer permission docs (#20102)
DataDog/documentation: content/en/serverless/guide/layer_not_authorized.md

---
title: Troubleshooting Serverless Layer not Authorized Errors
kind: documentation
---

> This guide helps you troubleshoot the deployment error `not authorized to perform: lambda:GetLayerVersion on resource`. This error is commonly seen with Datadog Library Layers or the Datadog Extension layer.

Reviewer: ```suggestion
This guide helps you troubleshoot the deployment error `not authorized to perform: lambda:GetLayerVersion on resource`. This error is commonly seen with Datadog Lambda Library layers or the Datadog Extension layer.
```

## Regionality

> Lambda functions can only include [Lambda Layers][1] which are located in the same region as the function. Typically this error occurs when users copy instrumentation settings from other applications deployed in different regions.

Reviewer: ```suggestion
Lambda functions can only include [Lambda layers][1] that are located in the same region as the function. Typically, this error occurs when users copy instrumentation settings from other applications deployed in different regions.
```

> Lambda functions can only included [Lambda Layers][1] which are located in the same region as the function. Typically this error occurs when users copy instrumentation settings from other applications deployed in different regions.

Reviewer 2: Small typo

```suggestion
Lambda functions can only include [Lambda Layers][1] which are located in the same region as the function. Typically this error occurs when users copy instrumentation settings from other applications deployed in different regions.
```

> Verify that the Lambda Layer region and Lambda function version match. Then verify that the version number is correct.

Reviewer: ```suggestion
Verify that the Lambda layer region and Lambda function version match. Then verify that the version number is correct.
```

> You can verify that a Lambda Layer version exists using any valid AWS credentials and running the following command:

Reviewer: ```suggestion
You can verify that a Lambda layer version exists by running `aws lambda get-layer-version` with valid AWS credentials.

For example, to verify the Datadog Extension layer, run:
```

```

> aws lambda get-layer-version \
>   --layer-name arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Node16-x \
>   --version-number {{< latest-lambda-layer-version layer="node" >}}
> 

Reviewer: ```suggestion
```

let's have one (and only one) example, or this will be confusing

Reviewer: That sounds good to me! Let's then explicitly say "For example, to verify the Datadog Extension layer and the Datadog Node.js library layer, run:" in line 14.

Author: Perfect, thank you!

aws lambda get-layer-version \
  --layer-name arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Extension \
  --version-number {{< latest-lambda-layer-version layer="extension" >}}

> ```

Reviewer 3: The example below from the package too large doc includes

> Replace the layer name and version number in the following example with those used by your applications:

I'm thinking we should add that line here too.

> ```

Author: My goal here is to list a couple of examples, I think that's okay because the error message will give them the exact layer name and version, so they can plug that into the command.

Happy to hear feedback on specific wording if you think that's confusing though. Thanks!

> ```

Reviewer: We don't need to list a command for every layer, but we need to be explicit that the command we give here is just an example. I'll make a suggestion to the wording above.

## Permissions

> Occasionally users accidentally explictly `DENY` permission for their functions to perform `lambda:GetLayerVersion`. Some [resource-based][2] policy configurations can cause an explicit `DENY`. Additionally IAM [Policy Boundaries][3] may also cause an explict `DENY` for `lambda:GetLayerVersion`.

Reviewer: ```suggestion
Occasionally, users accidentally explicitly `DENY` permission for their functions to perform `lambda:GetLayerVersion`. Some [resource-based][2] policy configurations can cause an explicit `DENY`. Additionally, IAM [permissions boundaries][3] may also cause an explicit `DENY` for `lambda:GetLayerVersion`.
```

> To test this, use an IAM user attached to the same IAM Policy your Lambda function uses, and test the `get-layer-version` command as listed above. The command should succeed with no errors.

Reviewer: ```suggestion
To test this, use an IAM user attached to the same IAM policy your Lambda function uses, and test the `get-layer-version` command as listed above. The command should succeed with no errors.
```

> ## Get help

Reviewer: ```suggestion
## Contact Datadog support
```

If you need the Datadog support team to help investigate, include the following information in your ticket:

1. The function's configured Lambda layers (name and version, or ARNs).

> 2. The project configuration files, with **redacted hardcoded secrets**: `serverless.yaml`, `package.json`, `package-lock.json`, `yarn.lock`, `tsconfig.json` and `webpack.config.json`.

Reviewer: ```suggestion
2. The project configuration files, with **redacted hardcoded secrets**: `serverless.yaml`, `package.json`, `package-lock.json`, `yarn.lock`, `tsconfig.json`, and `webpack.config.json`.
```
